Fix crashes/assertions when calling internals API on a window without frame

fred-wang opened the pull request:
#### e63fe9d14548d612e19da56efce3dab4f1a59ae2
<pre>
Fix crashes/assertions when calling internals API on a window without frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=286252">https://bugs.webkit.org/show_bug.cgi?id=286252</a>

Reviewed by Anne van Kesteren and Darin Adler.

There are multiple places in Internals.cpp that assumes a document, a
frame, a page or valid parameters ; and would otherwise crash or assert.
In this patch, we add a generic test case exercising Window.internals
properties when the associated frame is null. It calls the getter and
setter for each of these properties, or if the property is a function,
executes it with ten &apos;1&apos; as arguments. This should cover a large amount
of cases detectable by fuzzers. We also modify the code to fix the
issues found with that new test but there are probably more. Some of
these APIs should probably throw in case of failure, but we don&apos;t
necessarily modify their signature in this patch.

* LayoutTests/fast/harness/internals-object-property-access-on-window-without-frame-crash-expected.txt: Added.
* LayoutTests/fast/harness/internals-object-property-access-on-window-without-frame-crash.html: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setStrictRawResourceValidationPolicyDisabled): null-check frame.
(WebCore::Internals::numberOfActiveAnimations const): ditto.
(WebCore::Internals::numberOfAnimationTimelineInvalidations const): ditto.
(WebCore::Internals::insertTextPlaceholder): ditto.
(WebCore::Internals::formControlStateOfPreviousHistoryItem): ditto.
(WebCore::Internals::setFormControlStateOfPreviousHistoryItem): ditto.
(WebCore::Internals::testProcessIncomingSyncMessagesWhenWaitingForSyncReply): null-check document, frame and page and throw.
(WebCore::Internals::isDocumentAlive const): Replace ASSERT with an exception.
(WebCore::Internals::openDummyInspectorFrontend): null-check document, frame and page.
(WebCore::Internals::setInspectorIsUnderTest): ditto.
(WebCore::Internals::setFullscreenAutoHideDuration): ditto and replace ASSERT with an exception.
(WebCore::Internals::getReferencedFilePaths const): null-check frame.
(WebCore::Internals::startTrackingRenderingUpdates): null-check document, frame and page.
(WebCore::Internals::renderingUpdateCount): ditto.
(WebCore::Internals::setCompositingPolicyOverride): ditto.
(WebCore::Internals::compositingPolicyOverride const): ditto.
(WebCore::Internals::evaluateInWorldIgnoringException): null-check document and frame.
(WebCore::Internals::forceAXObjectCacheUpdate const): null-check document and axObjectCache.
(WebCore::Internals::forceReload): null-check frame.
(WebCore::Internals::reloadExpiredOnly): ditto.
(WebCore::Internals::setSelectionFromNone): ditto.
(WebCore::Internals::mockPaymentCoordinator): null-check page and throw.
(WebCore::Internals::systemColorForCSSValue): Replace RELEASE_ASSERT with an exception.
(WebCore::Internals::setTopDocumentURLForQuirks): null-check on frame and page.
* Source/WebCore/testing/Internals.h: Change some functions to return Exceptions.

Canonical link: <a href="https://commits.webkit.org/289309@main">https://commits.webkit.org/289309@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d245126b9a51a653d883e25c497e487d8d9f332d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86369 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40706 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91282 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37172 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13958 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66881 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24672 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89376 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4688 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78231 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47202 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4498 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32548 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36274 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75037 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33420 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93107 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13573 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9850 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75672 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13781 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74105 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74862 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18424 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19112 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17501 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13597 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18886 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13350 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16793 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15134 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -3953,10 +3963,10 @@ void Internals::setFullscreenInsets(FullscreenInsets insets)
-    ASSERT(page);
-
-    page->setFullscreenAutoHideDuration(Seconds(duration));
+    Document* document = contextDocument();
```
auto*, but prolly RefPtr?

annevk commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -2926,9 +2930,10 @@ void Internals::setAutomaticLinkDetectionEnabled(bool enabled)
-    ASSERT(contextDocument());
-    ASSERT(contextDocument()->page());
-    return contextDocument()->page()->chrome().client().testProcessIncomingSyncMessagesWhenWaitingForSyncReply();
+    Document* document = contextDocument();
```
auto* but prolly RefPtr

annevk commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1670,7 +1674,7 @@ ExceptionOr<Vector<AtomString>> Internals::formControlStateOfPreviousHistoryItem
 ExceptionOr<void> Internals::setFormControlStateOfPreviousHistoryItem(const Vector<AtomString>& state)
 {
-    HistoryItem* mainItem = frame()->loader().history().previousItem();
+    HistoryItem* mainItem = frame() ? frame()->loader().history().previousItem() : nullptr;
```
auto*

annevk commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1659,7 +1663,7 @@ void Internals::selectColorInColorChooser(HTMLInputElement& element, const Strin
 ExceptionOr<Vector<AtomString>> Internals::formControlStateOfPreviousHistoryItem()
 {
-    HistoryItem* mainItem = frame()->loader().history().previousItem();
+    HistoryItem* mainItem = frame() ? frame()->loader().history().previousItem() : nullptr;
```
auto*

annevk reviewed (approved):
Might be worth doing some RefPtr/Ref adoption while here.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1022,7 +1022,8 @@ void Internals::setOverrideResourceLoadPriority(ResourceLoadPriority priority)
 void Internals::setStrictRawResourceValidationPolicyDisabled(bool disabled)
 {
-    frame()->loader().setStrictRawResourceValidationPolicyDisabledForTesting(disabled);
+    if (auto* localFrame = frame())
```
I’d think this would need to be `RefPtr` instead of `auto*` in our new safer C++ model.

fred-wang replied:
For this case (and some others below) I think `loader()` and `setStrictRawResourceValidationPolicyDisabledForTesting()` are just trivial functions so that should be fine. But I agree let's try to use more smart pointers here, especially since the extra overhead is probably not important for these `internals` functions.

fred-wang replied:
Done.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1334,7 +1335,7 @@ bool Internals::animationWithIdExists(const String& id) const
 unsigned Internals::numberOfActiveAnimations() const
 {
-    return frame()->document()->timeline().numberOfActiveAnimationsForTesting();
+    return frame() ? frame()->document()->timeline().numberOfActiveAnimationsForTesting() : 0;
```
I’d think this would need to be in a `RefPtr` or call `protectedFrame()` in our new safer C++ model.

fred-wang replied:
Done.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1403,7 +1404,7 @@ Vector<Internals::AcceleratedAnimation> Internals::acceleratedAnimationsForEleme
 unsigned Internals::numberOfAnimationTimelineInvalidations() const
 {
-    return frame()->document()->timeline().numberOfAnimationTimelineInvalidationsForTesting();
+    return frame() ? frame()->document()->timeline().numberOfAnimationTimelineInvalidationsForTesting() : 0;
```
I’d think this would need to be in a `RefPtr` or call `protectedFrame()` in our new safer C++ model.

fred-wang replied:
Done.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1643,7 +1644,10 @@ void Internals::setCanShowPlaceholder(Element& element, bool canShowPlaceholder)
 Element* Internals::insertTextPlaceholder(int width, int height)
 {
-    return frame()->editor().insertTextPlaceholder(IntSize { width, height }).get();
+    auto* localFrame = frame();
```
I’d think this would need to be `RefPtr` instead of `auto*` in our new safer C++ model.

Not sure why this is using an `if` statement while the others are using `? :` instead.

fred-wang replied:
changed to `? :` style and to use RefPtr

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1659,7 +1663,7 @@ void Internals::selectColorInColorChooser(HTMLInputElement& element, const Strin
 ExceptionOr<Vector<AtomString>> Internals::formControlStateOfPreviousHistoryItem()
 {
-    HistoryItem* mainItem = frame()->loader().history().previousItem();
+    auto* mainItem = frame() ? frame()->loader().history().previousItem() : nullptr;
```
I’d think this would need to be `RefPtr` instead of `auto*` in our new safer C++ model.

fred-wang replied:
Done

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1670,7 +1674,7 @@ ExceptionOr<Vector<AtomString>> Internals::formControlStateOfPreviousHistoryItem
 ExceptionOr<void> Internals::setFormControlStateOfPreviousHistoryItem(const Vector<AtomString>& state)
 {
-    HistoryItem* mainItem = frame()->loader().history().previousItem();
+    auto* mainItem = frame() ? frame()->loader().history().previousItem() : nullptr;
```
I’d think this would need to be `RefPtr` instead of `auto*` in our new safer C++ model.

fred-wang replied:
Done

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -2924,11 +2928,12 @@ void Internals::setAutomaticLinkDetectionEnabled(bool enabled)
-    return contextDocument()->page()->chrome().client().testProcessIncomingSyncMessagesWhenWaitingForSyncReply();
+    RefPtr document = contextDocument();
+    if (!document || !document->page())
+        return Exception { ExceptionCode::InvalidStateError };
```
Is it important that this throw an exception rather than simply returning `false`?

fred-wang replied:
I'm not sure and don't have strong opinion on this. Currently, if we can't get the doc or page this would crash in debug mode so webkit developers can catch the issue. My idea was that throwing an exception would keep a similar hint that something bad is happening. If we just return false, probably it's harder to understand when such an issue is happening.

darinadler replied:
Crashes often come with a backtrace; exceptions do not do all that much to help someone identify what went wrong in a test. I suppose the main argument in favor of an exception is that just returning false might cause a false negative where a test would pass that should fail.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -3203,7 +3209,13 @@ String Internals::serviceWorkerClientInternalIdentifier(const Document& document
+    if (!contextDocument())
+        return nullptr;
+
+    auto* inspectedPage = contextDocument()->page();
```
I’d think this would need to be `RefPtr` instead of `auto*` in our new safer C++ model.

fred-wang replied:
OK, I changed this and other refcounted objects in this method (except frontendWindowProxy since WindowProxy::window() is trivial).

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -3203,7 +3209,13 @@ String Internals::serviceWorkerClientInternalIdentifier(const Document& document
 RefPtr<WindowProxy> Internals::openDummyInspectorFrontend(const String& url)
 {
-    auto* inspectedPage = contextDocument()->frame()->page();
+    if (!contextDocument())
```
Seems like this should go into a local variable the way it does in `setInspectorIsUnderTest`.

fred-wang replied:
Done

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -3224,11 +3236,11 @@ void Internals::closeDummyInspectorFrontend()
         return Exception { ExceptionCode::InvalidAccessError };
 
-    page->inspectorController().setIsUnderTest(isUnderTest);
+    document->page()->inspectorController().setIsUnderTest(isUnderTest);
```
I’d think this would need to be in a `RefPtr` or call `protectedPage()` in our new safer C++ model.

fred-wang replied:
Done

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -3951,12 +3963,13 @@ void Internals::setFullscreenInsets(FullscreenInsets insets)
+    RefPtr document = contextDocument();
+    if (!document || !document->page())
+        return Exception { ExceptionCode::InvalidStateError };
+    document->page()->setFullscreenAutoHideDuration(Seconds(duration));
```
I’d think this would need to be in a `RefPtr` or call `protectedFrame()` in our new safer C++ model.

fred-wang replied:
Done

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -4348,6 +4363,9 @@ bool Internals::isFromCurrentWorld(JSC::JSValue value) const
 {
     auto* document = contextDocument();
+    if (!document || !document->frame())
+        return JSC::JSValue { };
```
This is not right. We need to return something like `undefined` or `null` using a function like `jsUndefined()` or `jsNull()`, not `JSValue { }`.

fred-wang replied:
OK, it looks that I copied that from ScriptController::evaluateInWorldIgnoringException but indeed I meant to return null. Fixed.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -4362,8 +4380,10 @@ void Internals::setUsesOverlayScrollbars(bool enabled)
-    if (RefPtr document = contextDocument())
-        document->axObjectCache()->performDeferredCacheUpdate(ForceLayout::Yes);
+    if (RefPtr document = contextDocument()) {
+        if (auto* cache = document->axObjectCache())
```
I think this needs to be a smart pointer too for our new safer C++, not sure which one is right for `axObjectCache`.

fred-wang replied:
OK, it seems CheckedPtr is the one we should use for non-refcounted things: https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack

darinadler reviewed (changes requested):
The code that returns `JSValue { }` is wrong. Otherwise this is ready to go. I think it would be nice, but optional, to do a little better on the safer C++ idioms in all the code we are touching. But those are not wrong, just could be better and more consistent.

fred-wang committed:
Fix crashes/assertions when calling internals API on a window without frame
https://bugs.webkit.org/show_bug.cgi?id=286252

Reviewed by Anne van Kesteren and Darin Adler.

There are multiple places in Internals.cpp that assumes a document, a
frame, a page or valid parameters ; and would otherwise crash or assert.
In this patch, we add a generic test case exercising Window.internals
properties when the associated frame is null. It calls the getter and
setter for each of these properties, or if the property is a function,
executes it with ten '1' as arguments. This should cover a large amount
of cases detectable by fuzzers. We also modify the code to fix the
issues found with that new test but there are probably more. Some of
these APIs should probably throw in case of failure, but we don't
necessarily modify their signature in this patch.

* LayoutTests/fast/harness/internals-object-property-access-on-window-without-frame-crash-expected.txt: Added.
* LayoutTests/fast/harness/internals-object-property-access-on-window-without-frame-crash.html: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setStrictRawResourceValidationPolicyDisabled): null-check frame.
(WebCore::Internals::numberOfActiveAnimations const): ditto.
(WebCore::Internals::numberOfAnimationTimelineInvalidations const): ditto.
(WebCore::Internals::insertTextPlaceholder): ditto.
(WebCore::Internals::formControlStateOfPreviousHistoryItem): ditto.
(WebCore::Internals::setFormControlStateOfPreviousHistoryItem): ditto.
(WebCore::Internals::testProcessIncomingSyncMessagesWhenWaitingForSyncReply): null-check document, frame and page and throw.
(WebCore::Internals::isDocumentAlive const): Replace ASSERT with an exception.
(WebCore::Internals::openDummyInspectorFrontend): null-check document, frame and page.
(WebCore::Internals::setInspectorIsUnderTest): ditto.
(WebCore::Internals::setFullscreenAutoHideDuration): ditto and replace ASSERT with an exception.
(WebCore::Internals::getReferencedFilePaths const): null-check frame.
(WebCore::Internals::startTrackingRenderingUpdates): null-check document, frame and page.
(WebCore::Internals::renderingUpdateCount): ditto.
(WebCore::Internals::setCompositingPolicyOverride): ditto.
(WebCore::Internals::compositingPolicyOverride const): ditto.
(WebCore::Internals::evaluateInWorldIgnoringException): null-check document and frame.
(WebCore::Internals::forceAXObjectCacheUpdate const): null-check document and axObjectCache.
(WebCore::Internals::forceReload): null-check frame.
(WebCore::Internals::reloadExpiredOnly): ditto.
(WebCore::Internals::setSelectionFromNone): ditto.
(WebCore::Internals::mockPaymentCoordinator): null-check page and throw.
(WebCore::Internals::systemColorForCSSValue): Replace RELEASE_ASSERT with an exception.
(WebCore::Internals::setTopDocumentURLForQuirks): null-check on frame and page.
* Source/WebCore/testing/Internals.h: Change some functions to return Exceptions.

Canonical link: https://commits.webkit.org/289309@main
